grpcadapter/adapter.go

```diff
@@ -0,0 +1,770 @@
+	r.buf.Write(body)
+	return nil
+}
+
+func (r *bufferedEnvelopeReader) Read(data []byte) (int, error) {
+	if !r.hasInitialized {
+		r.initializedErr = r.fillBuffer()
+		r.hasInitialized = true
+	}
+	if r.initializedErr != nil {
+		return 0, r.initializedErr
+	}
+	return r.buf.Read(data)
+}
+
+func translateConnectCommonToGRPC(request *http.Request) {
+	request.ProtoMajor = 2
+	request.ProtoMinor = 0
+	header := request.Header
+	ensureTeTrailers(header)
+
+	delete(header, connectHeaderProtocolVersion)
+	delete(header, headerContentLength)
+
+	// Translate timeout header
+	if timeout := getHeaderCanonical(header, connectHeaderTimeout); len(timeout) > 0 {
+		delete(header, connectHeaderTimeout)
+		header[grpcHeaderTimeout] = []string{timeout + "m"}
+	}
+}
+
+func translateConnectStreamingToGRPC(request *http.Request, contentType string) {
+	translateConnectCommonToGRPC(request)
+	header := request.Header
+
+	grpcContentType := grpcContentTypePrefix + getTypeEncoding(contentType)
+	setHeaderCanonical(header, headerContentType, grpcContentType)
+	if contentEncoding := getHeaderCanonical(header, connectStreamingHeaderCompression); len(contentEncoding) > 0 {
+		delHeaderCanonical(header, connectStreamingHeaderCompression)
+		setHeaderCanonical(header, grpcHeaderCompression, contentEncoding)
+	}
+	if acceptEncoding := getHeaderCanonical(header, connectStreamingHeaderAcceptCompression); len(acceptEncoding) > 0 {
+		delHeaderCanonical(header, connectStreamingHeaderAcceptCompression)
+		setHeaderCanonical(header, grpcHeaderAcceptCompression, acceptEncoding)
+	}
+}
+
+func translateConnectUnaryToGRPC(request *http.Request, contentType string, readMaxBytes int) {
+	translateConnectCommonToGRPC(request)
+	header := request.Header
+
+	grpcContentType := grpcContentTypePrefix + strings.TrimPrefix(contentType, "application/")
+	setHeaderCanonical(header, headerContentType, grpcContentType)
+	if acceptEncoding := getHeaderCanonical(header, connectUnaryHeaderAcceptCompression); len(acceptEncoding) > 0 {
+		delHeaderCanonical(header, connectUnaryHeaderAcceptCompression)
+		setHeaderCanonical(header, grpcHeaderAcceptCompression, acceptEncoding)
+	}
+
+	// Handle GET request with query parameters
+	if request.Method == http.MethodGet {
+		request.Method = http.MethodPost
+		query := request.URL.Query()
+		request.URL.RawQuery = "" // clear query parameters
+
+		isCompressed := false
+		if contentEncoding := query.Get(connectUnaryCompressionQueryParameter); len(contentEncoding) > 0 {
+			isCompressed = contentEncoding != "identity"
+			delHeaderCanonical(header, connectUnaryHeaderCompression)
+			setHeaderCanonical(header, grpcHeaderCompression, contentEncoding)
+		}
+
+		msg := query.Get(connectUnaryMessageQueryParameter)
+		msgReader := queryValueReader(msg, query.Get(connectUnaryBase64QueryParameter) == "1")
+		request.Body = &bufferedEnvelopeReader{
+			ReadCloser:   io.NopCloser(msgReader),
+			isCompressed: isCompressed,
+			maxBytes:     readMaxBytes,
+		}
```

Comment:
Why do we need another buffer here? We ought to be able to calculate the payload size (or just eagerly un-base64 it and check the size).

Reply:
Mostly to move error handling to `io.Reader`. Simplifies path between GET and POST.